DESIGN.md

```diff
@@ -0,0 +1,756 @@
+to the kernel. The cancellation request will attempt to terminate the operation,
+causing it to complete immediately with an error. Cancellation is best-effort;
+the operation may or may not terminate early. If the operation does complete,
+the runtime discards the result. The runtime maintains the internal operation
+state until the completion as this state owns data the kernel may be
+referencing.
+
+## Read operations
+
+Depending on the flavor, read operation can have multiple io-uring operation
+representations. The io-uring API provides two different read operation codes:
+`IORING_OP_READ` and `IORING_OP_READ_FIXED`. The first accepts a buffer as a
+pointer, and the second takes a buffer as an identifier referencing a buffer
+pool and entry. Additionally, the `IORING_OP_READ` operation can accept a null
+buffer pointer, indicating that io-uring should pick a buffer from a provided
+buffer pool. The tokio-uring runtime will determine which opcode to use based on
+the kind of `IoBuf` kind it receives. The `Vec` kind maps to the `IORING_OP_READ`
+opcode, and the `Provided` maps to `IORING_OP_READ_FIXED`.
+
+## Prior art
+
+[Glommio] is an existing asynchronous Rust runtime built on top of io-uring.
+This proposal draws heavily from ideas presented there, tweaking concepts to
+line up with Tokio's idioms. [`@withoutboats`][boats] has explored the space
+with [ringbahn].
+
+The tokio-uring crate is built on the pure-rust [io-uring] crate, authored by
+[@quininer]. This crate provides a low-level interface to the io-uring syscalls.
+
+[Glommio]: https://github.com/DataDog/glommio
+[boats]: https://github.com/withoutboats
+[ringbahn]: https://github.com/ringbahn/ringbahn
+[@quininer]: https://github.com/quininer/
+[io-uring]: https://github.com/tokio-rs/io-uring/
+
+## Future work
+
+The main Tokio crate will likely adopts concepts from tokio-uring in the future.
+The most obvious area is Tokio's file system API, currently implemented using a
+thread-pool. The current API would remain, and the implementation would use
+io-uring when supported by the operating system. The tokio-uring APIs may form
+the basis for a Tokio 2.0 release, though this cannot happen until 2024 at the
+earliest.
+
+## Alternatives
+
+### Use a work-stealing scheduler
+
+An alternative approach could use a work-stealing scheduler, allowing
+underutilized worker threads to steal work from overloaded workers. While
+work-stealing is an efficient strategy for balancing load across worker threads,
+required synchronization adds overhead. An earlier article on the Tokio blog
+includes an overview of various scheduling strategies.
+
+The tokio-uring crate targets use-cases that can benefit from taking advantage
+of io-uring at the expense of discarding Tokio's portable API. These use cases
+will also benefit from reduced synchronization overhead and fine-grained control
+over thread load balancing strategies.
```

Comment:
I think not *requiring* synchronization in a sub-library is a good call. Those wanting finer grained control can require pinning to a thread. What about the possibility of making `tokio-uring` support `!Send` or `Send` generically, so that if someone *did* want to use a work-stealing multi-threaded scheduler, they can do so safely without wrapping the types in `Mutex`s? Or is that basically the best solution anyways?

Reply:
Odds are, `tokio-uring` will not be used when implementing a work-stealing backed scheduler on top. At that point, it would make more sense to add support for io-uring within tokio proper.